<a href="https://colab.research.google.com/github/irahanjose/Aprendizaje-de-M-quina-con-Python/blob/main/Parte0_Practica3_NoteBook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Antes de Empezar** 🛑

Este ejercicio puedes enviarlo para que podamos darte feedback. Entoces, **es necesario que hagas una copia** para que trabajes.


---

Cómo crear una copia:

*   En el menu, ve a *Archivo > Guardar una copia en Drive*

Cómo compartir el Colab:

*   Ve a Compartir, y después ingresa el correo electrónico alan@machinelearningenespanol.com. Haz clic en listo.

![](https://drive.google.com/uc?id=1H60wzNmRlflHd_cxHiRoSzogd5Khd42R)

---


**Iniciemos**

En este Colab, tienes que decidir qué modelo de aprendizaje usar dependiendo del problema y de los datos que tienes para ello.

Hecho con 💗 por alan@machinelearningenespanol.com

---
Hay varios modelos que podemos usar con sklearn. En la siguiente tabla, hay una lista de algunos. Puedes hacer clic sobre ellos para ir a la documentación, y aprender más.

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-c3ow{border-color:inherit;text-align:center;vertical-align:top}
.tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-0pky" colspan="2">Aprendizaje Supervisado</th>
    <th class="tg-0pky"></th>
    <th class="tg-0pky">Aprendizaje No Supervisado</th>
  </tr>
  <tr>
    <td class="tg-0pky">Clasificación</td>
    <td class="tg-c3ow">Regresión</td>
    <td class="tg-c3ow"></td>
    <td class="tg-c3ow">Agrupamiento</td>
  </tr>
  <tr>
    <td class="tg-0pky"><a href="https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html#sklearn.naive_bayes.GaussianNB">Näive Bayes</a></td>
    <td class="tg-c3ow"><a href="https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html">SVR</a></td>
    <td class="tg-c3ow"></td>
    <td class="tg-c3ow"><a href="https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html">K-medios</a></td>
  </tr>
  <tr>
    <td class="tg-0pky"><a href="https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html">SVC</a></td>
    <td class="tg-0pky"><a href="https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html">Decision Tree Regressor</a></td>
    <td class="tg-0pky"></td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-0pky"><a href="https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html">Decision Tree Classifier</a><br></td>
    <td class="tg-c3ow"></td>
    <td class="tg-c3ow"></td>
    <td class="tg-c3ow"></td>
  </tr>
</table>

Tu misión 🐱‍👤, si decides aceptarla, es: Elegir de la tabla de arriba un modelo adecuado para los siguientes problemas:

1. Una empresa de avalúos y asesoría inmobiliaria en Boston necesita una herramienta que le permita valuar viviendas 🏘️.
2. Un equipo de médicos necesita estimar el progreso de diabetes en un paciente después de un año de haber sido diagnosticado 🧪.
3. Un empresario vinícola está interesado en saber, en función del resultado de un análisis químico de vinos 🍷, de qué tres variedades diferentes en Italia se fermentó el producto.
4. Se requiere de un sistema que decida si una paciente tiene cáncer de mama o no. Este sistema ayudará a los médicos a decidir qué pacientes necesitan una valoración médica de un profesional de la salud 👩‍⚕️.

Para resolver el problema, usaremos los datos de juguete de sklearn. La información para cada dateset (significado de los atributos y etiquetas) la encuentras en:

1.   [Boston](https://scikit-learn.org/stable/datasets/toy_dataset.html#boston-dataset)
2.   [Breast Cancer](https://scikit-learn.org/stable/datasets/toy_dataset.html#breast-cancer-dataset)
3.   [Wine](https://scikit-learn.org/stable/datasets/toy_dataset.html#wine-dataset)
4.   [Diabetes](https://scikit-learn.org/stable/datasets/toy_dataset.html#diabetes-dataset)



**Dataset de viviendas**

Problema: Una empresa de avalúos y asesoría inmobiliaria en Boston necesita una herramienta que le permita valuar viviendas.

In [ ]:
import pandas as pd
from sklearn.datasets import load_boston
BostonDataset = load_boston()
df_Boston = pd.DataFrame(BostonDataset.data, columns=BostonDataset.feature_names)
df_Boston['target'] = pd.Series(BostonDataset.target)
df_Boston

CRIM    ZN  INDUS  CHAS    NOX  ...    TAX  PTRATIO       B  LSTAT  target
0    0.00632  18.0   2.31   0.0  0.538  ...  296.0     15.3  396.90   4.98    24.0
1    0.02731   0.0   7.07   0.0  0.469  ...  242.0     17.8  396.90   9.14    21.6
2    0.02729   0.0   7.07   0.0  0.469  ...  242.0     17.8  392.83   4.03    34.7
3    0.03237   0.0   2.18   0.0  0.458  ...  222.0     18.7  394.63   2.94    33.4
4    0.06905   0.0   2.18   0.0  0.458  ...  222.0     18.7  396.90   5.33    36.2
..       ...   ...    ...   ...    ...  ...    ...      ...     ...    ...     ...
501  0.06263   0.0  11.93   0.0  0.573  ...  273.0     21.0  391.99   9.67    22.4
502  0.04527   0.0  11.93   0.0  0.573  ...  273.0     21.0  396.90   9.08    20.6
503  0.06076   0.0  11.93   0.0  0.573  ...  273.0     21.0  396.90   5.64    23.9
504  0.10959   0.0  11.93   0.0  0.573  ...  273.0     21.0  393.45   6.48    22.0
505  0.04741   0.0  11.93   0.0  0.573  ...  273.0     21.0  396.90   7.88    11.9

[506 rows x 14 columns]

**Tu propuesta**

El problema pide generar una herramienta para valuar viviendas. Por lo general, la valoración de una vivienda se da como un precio, que es un valor específico. Entonces este suena a un problema de regresión. Tenemos que ver los datos para confirmar que podemos hacer una regresión.

Al ver los datos, tenemos que preguntarnos:

1.   ¿Contamos con etiquetas o target? Si las tenemos podemos usar aprendizaje supervisado. Si no tenemos etiquetas, es aprendizaje no supervisado.
2.   ¿De qué tipo es la etiqueta o target? ¿Es una clase discreta? ¿Es un valor contínuo específico? Si es una etiqueta de clase, entonces es clasificación. Si es una etiqueta o target continuo (valor específico), entonces es un problema de regresión.

Al ejecutar la celda de arriba, vemos el dataset y en la última columna nos damos cuenta que el target es un valor contínuo. Por ejemplo, para la instancia 0, el valor del target es de 24.0; y para la instancia 4, de 36.2. Así que confirmamos que es un problema de aprendizaje supervisado, y más específicamente, de regresión. También hemos ido a la documentación del dataset para ver la descripción de las columnas.

Sí el dataset hubiera contenido un target con solo tres opciones: barato, medio, caro. Entonces este sería un problema de clasificación, y nuestra herramienta para valuar sólo tendría estas tres opciones.

In [ ]:
X = df_Boston.iloc[:,0:-1] # Atributos: Todas las filas, de todas las columnas exceptuando la última
y = df_Boston.iloc[:,-1]   # Etiqueta: Última columna

# Entrenar el modelo
from sklearn.svm import SVR   # Importar
regr = SVR(C=1.0, epsilon=0.2)# Crear
regr.fit(X, y)                # Entrenar o ajustar

SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.2, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

**Dataset de pacientes con diabetes**

In [ ]:
import pandas as pd
from sklearn.datasets import load_diabetes
DiabetesDataset = load_diabetes()
df_Diabetes = pd.DataFrame(DiabetesDataset.data, columns=DiabetesDataset.feature_names)
df_Diabetes['target'] = pd.Series(DiabetesDataset.target)
df_Diabetes

age       sex       bmi  ...        s5        s6  target
0    0.038076  0.050680  0.061696  ...  0.019908 -0.017646   151.0
1   -0.001882 -0.044642 -0.051474  ... -0.068330 -0.092204    75.0
2    0.085299  0.050680  0.044451  ...  0.002864 -0.025930   141.0
3   -0.089063 -0.044642 -0.011595  ...  0.022692 -0.009362   206.0
4    0.005383 -0.044642 -0.036385  ... -0.031991 -0.046641   135.0
..        ...       ...       ...  ...       ...       ...     ...
437  0.041708  0.050680  0.019662  ...  0.031193  0.007207   178.0
438 -0.005515  0.050680 -0.015906  ... -0.018118  0.044485   104.0
439  0.041708  0.050680 -0.015906  ... -0.046879  0.015491   132.0
440 -0.045472 -0.044642  0.039062  ...  0.044528 -0.025930   220.0
441 -0.045472 -0.044642 -0.073030  ... -0.004220  0.003064    57.0

[442 rows x 11 columns]

**Tu propuesta**

**Dataset de análisis químico de vinos**

In [ ]:
import pandas as pd
from sklearn.datasets import load_wine
WineDataset = load_wine()
df_Wine = pd.DataFrame(WineDataset.data, columns=WineDataset.feature_names)
df_Wine['target'] = pd.Series(WineDataset.target)
df_Wine

alcohol  malic_acid   ash  ...  od280/od315_of_diluted_wines  proline  target
0      14.23        1.71  2.43  ...                          3.92   1065.0       0
1      13.20        1.78  2.14  ...                          3.40   1050.0       0
2      13.16        2.36  2.67  ...                          3.17   1185.0       0
3      14.37        1.95  2.50  ...                          3.45   1480.0       0
4      13.24        2.59  2.87  ...                          2.93    735.0       0
..       ...         ...   ...  ...                           ...      ...     ...
173    13.71        5.65  2.45  ...                          1.74    740.0       2
174    13.40        3.91  2.48  ...                          1.56    750.0       2
175    13.27        4.28  2.26  ...                          1.56    835.0       2
176    13.17        2.59  2.37  ...                          1.62    840.0       2
177    14.13        4.10  2.74  ...                          1.60    560.0       2

[178 rows x 14 columns]

**Tu propuesta**

**Dataset de pacientes con y sin cancer de mama**

In [ ]:
import pandas as pd
from sklearn.datasets import load_breast_cancer
BreastCancerDataset = load_breast_cancer()
df_BreastCancerDataset = pd.DataFrame(BreastCancerDataset.data, columns=BreastCancerDataset.feature_names)
df_BreastCancerDataset['target'] = pd.Series(BreastCancerDataset.target)
df_BreastCancerDataset

mean radius  mean texture  ...  worst fractal dimension  target
0          17.99         10.38  ...                  0.11890       0
1          20.57         17.77  ...                  0.08902       0
2          19.69         21.25  ...                  0.08758       0
3          11.42         20.38  ...                  0.17300       0
4          20.29         14.34  ...                  0.07678       0
..           ...           ...  ...                      ...     ...
564        21.56         22.39  ...                  0.07115       0
565        20.13         28.25  ...                  0.06637       0
566        16.60         28.08  ...                  0.07820       0
567        20.60         29.33  ...                  0.12400       0
568         7.76         24.54  ...                  0.07039       1

[569 rows x 31 columns]

**Tu propuesta**